### ISSUE TITLE
Bower packages are case sensitive

### ISSUE THREAD

**Issue by MaksimKiselev** [hiqdev/asset-packagist#26]
Please see https://github.com/fxpio/composer-asset-plugin/issues/70

asset-packagist also are case sensitive :worried:

**Comment by hiqsol** [comment]
Please show me the `composer.json` which is in trouble.

**Comment by MKiselev** [comment]
@hiqsol 
I was add in require section ```bower-asset/Sortable``` and expect that will be downloaded https://github.com/RubaXa/Sortable but asset-packagist was download ```bower-asset/sortable``` https://github.com/HubSpot/sortable

**Comment by githubjeka** [comment]
same here :(

```json
"require": {        
        "bower-asset/At.js": "^1.5.1",
    },
```

**Comment by SilverFire** [comment]
Seems to be annoying. Will try to dedicate some time and fix it

**Comment by SilverFire** [comment]
Composer is case-insensitive, so package name, listed in `require` section will be lowercased to install.

We can create aliases for packages, that are not lowercase to distinguish `bower-asset/At.js` from `bower-asset/at.js`. For example, we can add hash to the package name:

```php
$hash = substr(md5('bower-asset/At.js'), 0, 6); // 8fe67c
```

Then use this part of hash in the package name, for example `bower-asset/At.js` will be aliased to `bower-asset/at.js-8fe67c`.

Asset-packagist will save this alias and it's package representation and provide correct package.

The problem I see for now is that package will be saved locally in directory `vendor/bower-asset/at.js-8fe67c` so we need either to use [oomphinc/composer-installers-extender](https://github.com/oomphinc/composer-installers-extender) and map this directory back to `vendor/bower-asset/at.js` or use it as-is. In case we use it as-is, we create logical dependency on asset-packagist with the package name.

I can't see another way how we can overcome this problem, at least for now. In case anyone have better ideas - I will be glad to hear your thoughts.

@francoispluchino

**Comment by edgardmessias** [comment]
My idea: ASCII table

Example:
```
"bower-asset/At.js"    => "bower-asset/&#65;t.js"
"bower-asset/Sortable" => "bower-asset/&#83ortable"
```
I'm doing a local test.

**Comment by SilverFire** [comment]
It looks much less readable

**Comment by rob006** [comment]
We could use double underscore: `bower-asset/__at.js`. Or any special char, like `bower-asset/^at.js`.

**Comment by SilverFire** [comment]
Before each uppercase letter?

**Comment by rob006** [comment]
Yes.

**Comment by francoispluchino** [comment]
@SilverFire The problem will not be resolved for the dependencies of the root package. The Solver SAT names the package in lowercase before the search. Therefore, all Composer repositories receive a package name in lowercase.

Technically, nothing blocks to edit Composer and allow some packages to be case sensitive, But lead developers of Composer do not want to accept this change, it's all the more frustrating that there have not much change (search the function `strtolower()`). It was to 2 years ago, maybe it's changed now.

**Comment by SilverFire** [comment]
> The Solver SAT names the package in lowercase before the search. Therefore, all Composer repositories receive a package name in lowercase.

That's right, but your `composer-asset-plugin` fetches non-lowercase packages from Bower/NPM without any problem: https://asset-packagist.org/p/bower-asset/Blueprint-SlidePushMenus/latest.json

As it was suggested before, we can provide these packages in lowercase with mutated name as it was suggested above:
 - `https://asset-packagist.org/p/bower-asset/blueprint-slidepushmenus-a1b2c3/latest.json`: 
 - `https://asset-packagist.org/p/bower-asset/^blueprint-^slide^push^menus/latest.json`
 - `https://asset-packagist.org/p/bower-asset/__blueprint-__slide__push__menus/latest.json`

I don't thing Jordi will change something in Composer for this problem :)

**Comment by francoispluchino** [comment]
Yes it's the case, but the package name is normalized in Composer. So, if you have 2 packages with the same name but with a different case, Composer will consider that it is the same package in the resolution of dependencies. Therefore, it is not certain that we install the right package.

Non-exhaustive list of calls to the `strtolower` function which is problematic:

- `Composer\Command\ShowCommand` [L491](https://github.com/composer/composer/blob/master/src/Composer/Command/ShowCommand.php#L491)
- `Composer\Command\UpdateCommand` [L173](https://github.com/composer/composer/blob/master/src/Composer/Command/UpdateCommand.php#L173)
- `Composer\Command\UpdateCommand` [L195](https://github.com/composer/composer/blob/master/src/Composer/Command/UpdateCommand.php#L195)
- `Composer\DependencyResolver\Request` [L59](https://github.com/composer/composer/blob/master/src/Composer/DependencyResolver/Request.php#L59)
- `Composer\Package\BasePackage` [L69](https://github.com/composer/composer/blob/master/src/Composer/Package/BasePackage.php#L69)
- `Composer\Package\Link` [L60](https://github.com/composer/composer/blob/master/src/Composer/Package/Link.php#L60)
- `Composer\Package\Link` [L61](https://github.com/composer/composer/blob/master/src/Composer/Package/Link.php#L61)
- `Composer\Package\Loader\ArrayLoader` [L244](https://github.com/composer/composer/blob/master/src/Composer/Package/Loader/ArrayLoader.php#L244)
- `Composer\Repository\ArrayRepository` [L44](https://github.com/composer/composer/blob/master/src/Composer/Repository\ArrayRepository.php#L44)
- `Composer\Repository\ArrayRepository` [L69](https://github.com/composer/composer/blob/master/src/Composer/Repository\ArrayRepository.php#L69)
- `Composer\Repository\ComposerRepository` [L125](https://github.com/composer/composer/blob/master/src/Composer/Repository\ComposerRepository.php#L125)
- `Composer\Repository\ComposerRepository` [L156](https://github.com/composer/composer/blob/master/src/Composer/Repository\ComposerRepository.php#L156)
- `Composer\Repository\ComposerRepository` [L380](https://github.com/composer/composer/blob/master/src/Composer/Repository\ComposerRepository.php#L380)
- `Composer\Repository\ComposerRepository` [L798](https://github.com/composer/composer/blob/master/src/Composer/Repository\ComposerRepository.php#L798)
- `Composer\Repository\ComposerRepository` [L803](https://github.com/composer/composer/blob/master/src/Composer/Repository\ComposerRepository.php#L803)
- `Composer\Repository\ComposerRepository` [L808](https://github.com/composer/composer/blob/master/src/Composer/Repository\ComposerRepository.php#L808)

PS. It was mainly Nils who did not want to modify the Solver SAT.

**Comment by rob006** [comment]
@francoispluchino But if in `require` section of my `composer.json` I will use `bower-asset/^at.js` instead of `bower-asset/At.js` are there any technical problems with treating `bower-asset/^at.js` as `At.js` bower package? For composer it will be `bower-asset/^at.js` so `strtolower` is irrelevant.  It will not be 100% transparent to users, but at least it works.

**Comment by francoispluchino** [comment]
@rob006 Why not! If you have a solution, can you do a PR on [fxpio/composer-asset-plugin](https://github.com/fxpio/composer-asset-plugin)? Did you check if the special characters are accepted by Composer?

**Comment by rob006** [comment]
> Did you check if the special characters are accepted by Composer?

According to [docs](https://getcomposer.org/doc/04-schema.md#name) there are no limitations, only recommendations. Bower [does not allow `^` in package name](https://github.com/bower/spec/blob/master/json.md#name), so there should be no conflicts (at least in theory).

Unfortunately, I don't have time to implement it myself.

**Comment by edgardmessias** [comment]
I started working

**Comment by totten** [comment]
It *feels* like uppercase characters aren't very common in bower package names -- but the numbers matter. If 20% of packages have uppercase, then it's a big problem. If 0.5% of packages, then downstream folks could look for workarounds (eg don't use a particular asset, or download it manually, or ask upstream to rename).

To get a  ballpark, I wrote a [quick and dirty script](https://gist.github.com/totten/bd439f01adc46cb8c344ca6dc4b2783a). Overall,  ~6% of bower packages seem to have an uppercase letter. The distribution varies a bit depending on what kind of packages you look at, e.g.
 * _Above-average prevalence_: jQuery (10%) and Backbone(7%)
 * _Below-average prevalence_: Angular (2%), Ember (2%), React (1%), Bootstrap (1%)

**Comment by hiqsol** [comment]
I have a simple but a controversial fix #64 
Please, see and comment.
